In [2]:
import sys
import pandas as pd
import numpy as np
import matplotlib

print("Python:", sys.version)
print("Pandas:", pd.__version__)
print("NumPy:", np.__version__)
print("Matplotlib:", matplotlib.__version__)

Python: 3.13.14 (v3.13.14:fd17997c386, Jun 10 2026, 08:55:00) [Clang 21.0.0 (clang-2100.1.1.101)]
Pandas: 3.0.5
NumPy: 2.5.1
Matplotlib: 3.11.1


In [3]:
import sys
import pandas as pd
import numpy as np
import matplotlib

print("Python:", sys.version)
print("Pandas:", pd.__version__)
print("NumPy:", np.__version__)
print("Matplotlib:", matplotlib.__version__)

Python: 3.13.14 (v3.13.14:fd17997c386, Jun 10 2026, 08:55:00) [Clang 21.0.0 (clang-2100.1.1.101)]
Pandas: 3.0.5
NumPy: 2.5.1
Matplotlib: 3.11.1


In [4]:
import pandas as pd

file_path = "../data/raw/sample_-_superstore.xls"

df = pd.read_excel(file_path)

df.head()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country/Region,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,US-2023-103800,2023-01-03,2023-01-07,Standard Class,DP-13000,Darren Powers,Consumer,United States,Houston,...,77095,Central,OFF-PA-10000174,Office Supplies,Paper,"Message Book, Wirebound, Four 5 1/2"" X 4"" Form...",16.448,2,0.2,5.5512
1,2,US-2023-112326,2023-01-04,2023-01-08,Standard Class,PO-19195,Phillina Ober,Home Office,United States,Naperville,...,60540,Central,OFF-BI-10004094,Office Supplies,Binders,GBC Standard Plastic Binding Systems Combs,3.540,2,0.8,-5.4870
2,3,US-2023-112326,2023-01-04,2023-01-08,Standard Class,PO-19195,Phillina Ober,Home Office,United States,Naperville,...,60540,Central,OFF-LA-10003223,Office Supplies,Labels,Avery 508,11.784,3,0.2,4.2717
3,4,US-2023-112326,2023-01-04,2023-01-08,Standard Class,PO-19195,Phillina Ober,Home Office,United States,Naperville,...,60540,Central,OFF-ST-10002743,Office Supplies,Storage,SAFCO Boltless Steel Shelving,272.736,3,0.2,-64.7748
4,5,US-2023-141817,2023-01-05,2023-01-12,Standard Class,MB-18085,Mick Brown,Consumer,United States,Philadelphia,...,19143,East,OFF-AR-10003478,Office Supplies,Art,Avery Hi-Liter EverBold Pen Style Fluorescent ...,19.536,3,0.2,4.8840


In [5]:
import pandas as pd

file_path = "../data/raw/sample_-_superstore.xls"

df = pd.read_excel(file_path)

df.head()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country/Region,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,US-2023-103800,2023-01-03,2023-01-07,Standard Class,DP-13000,Darren Powers,Consumer,United States,Houston,...,77095,Central,OFF-PA-10000174,Office Supplies,Paper,"Message Book, Wirebound, Four 5 1/2"" X 4"" Form...",16.448,2,0.2,5.5512
1,2,US-2023-112326,2023-01-04,2023-01-08,Standard Class,PO-19195,Phillina Ober,Home Office,United States,Naperville,...,60540,Central,OFF-BI-10004094,Office Supplies,Binders,GBC Standard Plastic Binding Systems Combs,3.540,2,0.8,-5.4870
2,3,US-2023-112326,2023-01-04,2023-01-08,Standard Class,PO-19195,Phillina Ober,Home Office,United States,Naperville,...,60540,Central,OFF-LA-10003223,Office Supplies,Labels,Avery 508,11.784,3,0.2,4.2717
3,4,US-2023-112326,2023-01-04,2023-01-08,Standard Class,PO-19195,Phillina Ober,Home Office,United States,Naperville,...,60540,Central,OFF-ST-10002743,Office Supplies,Storage,SAFCO Boltless Steel Shelving,272.736,3,0.2,-64.7748
4,5,US-2023-141817,2023-01-05,2023-01-12,Standard Class,MB-18085,Mick Brown,Consumer,United States,Philadelphia,...,19143,East,OFF-AR-10003478,Office Supplies,Art,Avery Hi-Liter EverBold Pen Style Fluorescent ...,19.536,3,0.2,4.8840


In [6]:
print(f"Rows: {df.shape[0]}")
print(f"Columns: {df.shape[1]}")

display(df.info())

display(df.columns)

Rows: 10194
Columns: 21
<class 'pandas.DataFrame'>
RangeIndex: 10194 entries, 0 to 10193
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   Row ID          10194 non-null  int64         
 1   Order ID        10194 non-null  str           
 2   Order Date      10194 non-null  datetime64[us]
 3   Ship Date       10194 non-null  datetime64[us]
 4   Ship Mode       10194 non-null  str           
 5   Customer ID     10194 non-null  str           
 6   Customer Name   10194 non-null  str           
 7   Segment         10194 non-null  str           
 8   Country/Region  10194 non-null  str           
 9   City            10194 non-null  str           
 10  State/Province  10194 non-null  str           
 11  Postal Code     10194 non-null  object        
 12  Region          10194 non-null  str           
 13  Product ID      10194 non-null  str           
 14  Category        10194 non-null  str      

None

Index(['Row ID', 'Order ID', 'Order Date', 'Ship Date', 'Ship Mode',
       'Customer ID', 'Customer Name', 'Segment', 'Country/Region', 'City',
       'State/Province', 'Postal Code', 'Region', 'Product ID', 'Category',
       'Sub-Category', 'Product Name', 'Sales', 'Quantity', 'Discount',
       'Profit'],
      dtype='str')

## Data Validation Summary

The Sample Superstore dataset loaded successfully.

Initial validation indicates:

- 10,194 records
- 21 columns
- No missing values
- Date columns correctly recognized as datetime
- Sales, Profit, and Discount correctly recognized as numeric
- Dataset appears suitable for exploratory analysis

Next step:
Validate duplicate records and confirm business data quality before performing analysis.

In [7]:
duplicate_rows = df.duplicated().sum()

print(f"Duplicate rows: {duplicate_rows}")

Duplicate rows: 0


Additional validation:

- No duplicate records identified.
- Dataset passed initial data quality validation.
- Approved for exploratory data analysis (EDA).

## Data Cleaning

The dataset passed initial validation with no missing values or exact duplicate records.

This stage evaluates data consistency and business-rule validity before analysis.

Cleaning changes will only be made where a documented data-quality issue is identified.

In [8]:
print("Sales <= 0:", (df["Sales"] <= 0).sum())
print("Quantity <= 0:", (df["Quantity"] <= 0).sum())
print("Discount < 0:", (df["Discount"] < 0).sum())
print("Discount > 1:", (df["Discount"] > 1).sum())
print("Profit < 0:", (df["Profit"] < 0).sum())

Sales <= 0: 0
Quantity <= 0: 0
Discount < 0: 0
Discount > 1: 0
Profit < 0: 1901


In [9]:
for column in ["Segment", "Region", "Category", "Ship Mode"]:
    print(f"\n{column}:")
    print(df[column].value_counts())


Segment:
Segment
Consumer       5281
Corporate      3090
Home Office    1823
Name: count, dtype: int64

Region:
Region
West       3253
East       2986
Central    2335
South      1620
Name: count, dtype: int64

Category:
Category
Office Supplies    6128
Furniture          2201
Technology         1865
Name: count, dtype: int64

Ship Mode:
Ship Mode
Standard Class    6120
Second Class      1979
First Class       1548
Same Day           547
Name: count, dtype: int64


In [10]:
print("Order date range:")
print(df["Order Date"].min(), "to", df["Order Date"].max())

print("\nShip date range:")
print(df["Ship Date"].min(), "to", df["Ship Date"].max())

invalid_shipping = (df["Ship Date"] < df["Order Date"]).sum()
print("\nShip dates before order dates:", invalid_shipping)

Order date range:
2023-01-03 00:00:00 to 2026-12-30 00:00:00

Ship date range:
2023-01-07 00:00:00 to 2027-01-05 00:00:00

Ship dates before order dates: 0


## Data Cleaning Summary

The dataset was evaluated for common data-quality and business-rule issues.

Validation confirmed:

- No missing values
- No exact duplicate records
- No nonpositive sales values
- No nonpositive quantity values
- No discounts outside the expected 0–1 range
- No shipping dates occurring before order dates
- Key categorical fields were consistently represented
- Order dates range from January 3, 2023 to December 30, 2026
- Ship dates range from January 7, 2023 to January 5, 2027
- 1,901 transactions have negative profit

Negative-profit transactions were retained because they represent potentially valid business outcomes rather than data-quality errors.

No corrective cleaning was required based on the validation performed. The original data will therefore be preserved for analysis.

## Exploratory Business Analysis

This stage examines sales performance, profitability, customer and product segments, geographic performance, and trends over time.

The objective is to identify the primary drivers of business performance and develop evidence-based recommendations.

In [11]:
total_sales = df["Sales"].sum()
total_profit = df["Profit"].sum()
total_orders = df["Order ID"].nunique()
total_quantity = df["Quantity"].sum()
profit_margin = total_profit / total_sales

print(f"Total Sales: ${total_sales:,.2f}")
print(f"Total Profit: ${total_profit:,.2f}")
print(f"Total Orders: {total_orders:,}")
print(f"Units Sold: {total_quantity:,}")
print(f"Overall Profit Margin: {profit_margin:.2%}")

Total Sales: $2,326,534.35
Total Profit: $292,296.81
Total Orders: 5,111
Units Sold: 38,654
Overall Profit Margin: 12.56%


In [12]:
category_performance = (
    df.groupby("Category")
      .agg(
          Sales=("Sales", "sum"),
          Profit=("Profit", "sum")
      )
      .sort_values("Sales", ascending=False)
)

category_performance["Profit Margin"] = (
    category_performance["Profit"] / category_performance["Sales"]
)

category_performance

,Sales,Profit,Profit Margin
Category,,,
Technology,839893.2790,146543.3756,0.174479
Furniture,754747.7613,19729.9956,0.026141
Office Supplies,731893.3140,126023.4434,0.172188


In [13]:
subcategory_performance = (
    df.groupby(["Category", "Sub-Category"])
      .agg(
          Sales=("Sales", "sum"),
          Profit=("Profit", "sum")
      )
)

subcategory_performance["Profit Margin"] = (
    subcategory_performance["Profit"] /
    subcategory_performance["Sales"]
)

subcategory_performance = (
    subcategory_performance
    .sort_values("Profit", ascending=True)
)

subcategory_performance

Sales      Profit  Profit Margin
Category        Sub-Category                                        
Furniture       Tables        208020.1820 -17753.2061      -0.085344
                Bookcases     115361.2043  -3632.0736      -0.031484
Office Supplies Supplies       46725.4980  -1171.3945      -0.025070
                Fasteners       8532.2400   2428.6358       0.284642
Technology      Machines      189925.0310   3461.9769       0.018228
Office Supplies Labels         12695.0420   5572.7780       0.438973
                Art            27659.0140   6653.1962       0.240544
                Envelopes      16528.3620   6988.0247       0.422790
Furniture       Furnishings    95598.1260  13891.7430       0.145314
Office Supplies Appliances    108213.1850  18329.4844       0.169383
                Storage       224644.5540  21285.1115       0.094750
Furniture       Chairs        335768.2490  27223.5323       0.081078
Office Supplies Binders       207354.8810  31426.1003       0.151557
                Paper          79540.5380  34511.5070       0.433886
Technology      Accessories   167380.3180  41936.6357       0.250547
                Phones        331842.6400  45050.8265       0.135760
                Copiers       150745.2900  56093.9365       0.372111

In [14]:
discount_analysis = (
    df.groupby("Discount")
      .agg(
          Sales=("Sales", "sum"),
          Profit=("Profit", "sum"),
          Transactions=("Order ID", "count")
      )
      .reset_index()
)

discount_analysis["Profit Margin"] = (
    discount_analysis["Profit"] /
    discount_analysis["Sales"]
)

discount_analysis

,Discount,Sales,Profit,Transactions,Profit Margin
0,0.00,1.105324e+06,326718.5872,4925,0.295586
1,0.10,5.495250e+04,9099.9700,96,0.165597
2,0.15,2.755852e+04,1418.9915,52,0.051490
3,0.20,7.739394e+05,91079.9517,3706,0.117684
4,0.30,1.044741e+05,-10513.4456,230,-0.100632
5,0.32,1.449346e+04,-2391.1377,27,-0.164980
6,0.40,1.164978e+05,-23086.3742,207,-0.198170
7,0.45,5.484974e+03,-2493.1111,11,-0.454535
8,0.50,5.891854e+04,-20506.4281,66,-0.348047
9,0.60,7.105940e+03,-6164.3472,149,-0.867492


In [15]:
furniture_discount_analysis = (
    df[df["Category"] == "Furniture"]
    .groupby(["Sub-Category", "Discount"])
    .agg(
        Sales=("Sales", "sum"),
        Profit=("Profit", "sum"),
        Transactions=("Order ID", "count")
    )
    .reset_index()
)

furniture_discount_analysis["Profit Margin"] = (
    furniture_discount_analysis["Profit"] /
    furniture_discount_analysis["Sales"]
)

furniture_discount_analysis[
    furniture_discount_analysis["Sub-Category"].isin(["Tables", "Bookcases"])
]

,Sub-Category,Discount,Sales,Profit,Transactions,Profit Margin
0,Bookcases,0.00,31935.9800,6075.7117,60,0.190247
1,Bookcases,0.15,27558.5215,1418.9915,52,0.051490
2,Bookcases,0.20,27178.1160,168.5882,48,0.006203
3,Bookcases,0.30,4282.6980,-555.8726,10,-0.129795
4,Bookcases,0.32,14493.4588,-2391.1377,27,-0.164980
5,Bookcases,0.50,7308.4600,-4255.8117,18,-0.582313
6,Bookcases,0.70,2603.9700,-4092.5430,17,-1.571655
15,Tables,0.00,71706.2300,13334.9247,76,0.185966
16,Tables,0.20,45636.5820,-298.3980,72,-0.006539
17,Tables,0.30,25902.9800,-3493.8376,56,-0.134882


### Profitability and Discount Finding

Furniture generated substantial sales but significantly weaker profitability than Technology and Office Supplies.

Sub-category analysis identified Tables and Bookcases as major contributors to the profitability weakness. Both generated negative profit overall.

Further analysis identified a strong association between discount levels and profitability:

- Undiscounted Bookcases generated approximately a 19.0% profit margin.
- Undiscounted Tables generated approximately an 18.6% profit margin.
- Bookcase margins became negative at discounts of 30% and above.
- Table margins were approximately break-even at a 20% discount and became increasingly negative at higher discount levels.
- Across the full dataset, every observed discount level of 30% or greater produced a negative aggregate profit margin.

These findings suggest that discount strategy is an important contributor to the profitability problem, particularly for Tables and Bookcases.

**Business implication:** Management should review discount policies for these sub-categories, particularly discounts of 30% or greater, and evaluate whether lower discount thresholds could preserve sales while improving profitability.

This analysis identifies an association between discounting and profitability and does not establish causation.

In [1]:
yearly_performance = (
    df.assign(Year=df["Order Date"].dt.year)
      .groupby("Year")
      .agg(
          Sales=("Sales", "sum"),
          Profit=("Profit", "sum"),
          Orders=("Order ID", "nunique")
      )
)

yearly_performance["Profit Margin"] = (
    yearly_performance["Profit"] /
    yearly_performance["Sales"]
)

yearly_performance

NameError: name 'df' is not defined

In [2]:
import pandas as pd

file_path = "../data/raw/sample_-_superstore.xls"

df = pd.read_excel(file_path)

print("Dataset loaded successfully.")
print("Rows:", len(df))
print("Columns:", len(df.columns))

Dataset loaded successfully.
Rows: 10194
Columns: 21


In [3]:
yearly_performance = (
    df.assign(Year=df["Order Date"].dt.year)
      .groupby("Year")
      .agg(
          Sales=("Sales", "sum"),
          Profit=("Profit", "sum"),
          Orders=("Order ID", "nunique")
      )
)

yearly_performance["Profit Margin"] = (
    yearly_performance["Profit"] /
    yearly_performance["Sales"]
)

yearly_performance

,Sales,Profit,Orders,Profit Margin
Year,,,,
2023,494040.2121,51684.2957,995,0.104616
2024,472993.0310,62020.9695,1053,0.131124
2025,613933.5800,82665.2018,1340,0.134648
2026,745567.5312,95926.3476,1723,0.128662


### Yearly Performance Finding

Sales, profit, and order volume show substantial overall growth from 2023 through 2026.

Sales increased from approximately $494K in 2023 to $746K in 2026, while orders increased from 995 to 1,723.

Profit margin improved from 10.46% in 2023 to 13.46% in 2025. However, margin declined to 12.87% in 2026 despite 2026 producing the highest sales and profit.

This suggests that revenue growth alone does not fully explain business performance. Further analysis is required to identify the factors contributing to the 2026 margin decline.

In [4]:
category_yearly = (
    df.assign(Year=df["Order Date"].dt.year)
      .groupby(["Year", "Category"])
      .agg(
          Sales=("Sales", "sum"),
          Profit=("Profit", "sum")
      )
      .reset_index()
)

category_yearly["Profit Margin"] = (
    category_yearly["Profit"] /
    category_yearly["Sales"]
)

category_yearly[
    category_yearly["Year"].isin([2025, 2026])
]

,Year,Category,Sales,Profit,Profit Margin
6,2025,Furniture,200979.7260,7216.2227,0.035905
7,2025,Office Supplies,186373.6940,35613.2171,0.191085
8,2025,Technology,226580.1600,39835.7620,0.175813
9,2026,Furniture,219312.9292,3349.2393,0.015272
10,2026,Office Supplies,253835.4350,41743.0242,0.164449
11,2026,Technology,272419.1670,50834.0841,0.186602


In [5]:
furniture_detail = (
    df[
        (df["Order Date"].dt.year.isin([2025, 2026])) &
        (df["Category"] == "Furniture")
    ]
    .assign(Year=lambda x: x["Order Date"].dt.year)
    .groupby(["Year", "Sub-Category"])
    .agg(
        Sales=("Sales", "sum"),
        Profit=("Profit", "sum")
    )
    .reset_index()
)

furniture_detail["Profit Margin"] = (
    furniture_detail["Profit"] /
    furniture_detail["Sales"]
)

furniture_detail

,Year,Sub-Category,Sales,Profit,Profit Margin
0,2025,Bookcases,26426.8805,127.5152,0.004825
1,2025,Chairs,85079.4310,5948.3844,0.069916
2,2025,Furnishings,28638.0040,4090.6649,0.142840
3,2025,Tables,60835.4105,-2950.3418,-0.048497
4,2026,Bookcases,30354.0737,-658.1879,-0.021684
5,2026,Chairs,98032.3770,7788.1024,0.079444
6,2026,Furnishings,29927.2760,4311.5985,0.144069
7,2026,Tables,60999.2025,-8092.2737,-0.132662


In [6]:
### 2026 Margin Decline Investigation

The decline in overall profit margin from 13.46% in 2025 to 12.87% in 2026 was not distributed evenly across the business.

Furniture margin declined from 3.59% to 1.53%, while Office Supplies also declined from 19.11% to 16.44%. Technology improved from 17.58% to 18.66%.

Within Furniture, Tables were the clearest source of deterioration. Table sales remained approximately $61K in both years, but losses increased from about $3.0K in 2025 to $8.1K in 2026. Table profit margin consequently declined from -4.85% to -13.27%.

Bookcases also moved from approximately break-even to a negative margin, while Chairs and Furnishings improved.

These results indicate that the Furniture decline was concentrated primarily in Tables rather than reflecting deterioration across all Furniture sub-categories.

SyntaxError: invalid decimal literal (3515296817.py, line 7)

### 2026 Margin Decline Investigation

The decline in overall profit margin from 13.46% in 2025 to 12.87% in 2026 was not distributed evenly across the business.

Furniture margin declined from 3.59% to 1.53%, while Office Supplies also declined from 19.11% to 16.44%. Technology improved from 17.58% to 18.66%.

Within Furniture, Tables were the clearest source of deterioration. Table sales remained approximately $61K in both years, but losses increased from about $3.0K in 2025 to $8.1K in 2026. Table profit margin consequently declined from -4.85% to -13.27%.

Bookcases also moved from approximately break-even to a negative margin, while Chairs and Furnishings improved.

These results indicate that the Furniture decline was concentrated primarily in Tables rather than reflecting deterioration across all Furniture sub-categories.